pkg/registry/autoscaling/horizontalpodautoscaler/strategy.go

```diff
@@ -69,10 +76,24 @@ func (autoscalerStrategy) AllowCreateOnUpdate() bool {
 func (autoscalerStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
 	newHPA := obj.(*autoscaling.HorizontalPodAutoscaler)
 	oldHPA := old.(*autoscaling.HorizontalPodAutoscaler)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HPAContainerMetrics) {
+		metrics := filterContainerMetricSource(newHPA.Spec.Metrics)
+		newHPA.Spec.Metrics = metrics
+	}
 	// Update is not allowed to set status
 	newHPA.Status = oldHPA.Status
 }
 
+func filterContainerMetricSource(metrics []autoscaling.MetricSpec) []autoscaling.MetricSpec {
```

Comment:
I'm assuming that the controller will have to handle the case where a metric source types object is present but the metric source definition is nil. Currently the controller does not do so.

Reply:
So it looks like after nulling out the container source it will be caught in the validate step which would throw an error. Please ignore my last comment.